[DirectX] Eliminate resource global variables from module

pow2clk opened the pull request:
By giving these intrinsics their appropriate attributes, loads of globals that are stored on the other side of these calls can be eliminated by the EarlyCSE pass.  Stores to the same globals and the globals themselves require more direct intervention as part of the handleFromBinding lowering.

Adds a test that verifies that the unneeded globals and their uses can be eliminated and also that the attributes are set properly.

Fixes #104271

Commit:
[DirectX] Mark buffer load/store as mem read/write

By giving these intrinsics their appropriate attributes, loads of allocas that are stored on the other side of these calls can be eliminated.

Adds a test that verifies that the unneeded loads can be eliminated and also that the attributes are set properly.

Fixes #104271

This may be the first part of a broader audit of

hekota commented:
This is certainly a step in the right direction, but it will probably not eliminate extra `load`a on a handle after a `typedBufferStore` is used on the same buffer, right? Or is that part of the "broader audit" you mentioned above?

Commit:
[DirectX] Eliminate resource global variables from module

By giving these intrinsics their appropriate attributes, loads of globals that are stored on the other side of these calls can be eliminated by the EarlyCSE pass.  Stores to the same globals and the globals themselves require more direct intervention as part of the handleFromBinding lowering.

Adds a test that verifies that the unneeded globals and their uses can be eliminated and also that the attributes are set properly.

Fixes #104271

coopp commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,93 @@
+; RUN: opt -S -passes='early-cse<memssa>' %s -o %t
+; RUN: FileCheck --check-prefixes=CSE,CHECK %s < %t
+; finish compiling to verify that dxil-op-lower removes the globals entirely
+; RUN: llc -mtriple=dxil-pc-shadermodel6.0-compute  --filetype=asm -o - %t | FileCheck --check-prefixes=LLC,CHECK %s
```
Why are both shader model 6.0 and 6.6 used here?  Is there some special difference?

pow2clk replied:
Yes, the difference is whether the handle is created or created and annotated. We have separate paths for them, so even though they call the same utility function, I thought it better to test both paths.

coopp commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,93 @@
+; Also that DXILOpLowering eliminates the globals entirely.
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxilv1.6-unknown-shadermodel6.6-compute"
```
Does this triple get overridden by the -mtriple set above in the tests?

pow2clk replied:
I'd rather be able to run both options above, I've removed this too.

coopp commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -204,6 +204,22 @@ class OpLowerer {
+  void removeResourceGlobals(CallInst *CI) {
+    for (User *User : make_early_inc_range(CI->users())) {
+      if(StoreInst *Store = dyn_cast<StoreInst>(User)) {
+	if (GlobalVariable *GV = dyn_cast<GlobalVariable>(Store->getOperand(1))) {
```
What is special about operand 1?  in `Store->getOperand(1)`?

pow2clk replied:
That's the location of the global that gets stored to. In DXC, we would declare constants that made clear what these refer to, but I see no similar practice in LLVM.

Commit:
Make GV removal more robust add opt dxil-op-lower test

Commit:
clang-format

pow2clk commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,94 @@
+; Ensure that EarlyCSE is able to eliminate unneeded loads of resource globals across typedBufferLoad.
+; Also that DXILOpLowering eliminates the globals entirely.
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
```
Removed

pow2clk commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,94 @@
+@In = global %"class.hlsl::RWBuffer" zeroinitializer, align 4
+@Out = global %"class.hlsl::RWBuffer" zeroinitializer, align 4
+
+; Function Attrs: convergent noinline norecurse
```
Okay, removed.

pow2clk commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,94 @@
+
+; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn
+; CSE: declare <4 x float> @llvm.dx.typedBufferLoad.v4f32.tdx.TypedBuffer_v4f32_1_0_0t(target("dx.TypedBuffer", <4 x float>, 1, 0, 0), i32) [[ROAttr:#[0-9]+]]
+declare <4 x float> @llvm.dx.typedBufferLoad.v4f32.tdx.TypedBuffer_v4f32_1_0_0t(target("dx.TypedBuffer", <4 x float>, 1, 0, 0), i32) #2
```
It's the only way I have at the moment to directly test the attributes applied in IntrinsicsDirectX.td. That does go better with a CodeGen test when these can be generated. I'm happy to leave that for that change and limit this testing to the indirect effects of the attribute here.

bogner replied:
So the LLVM IR parser has a feature where if you use an intrinsic without declaring it, it will find the declaration implicitly. So what I'm saying is that you could omit the `declare <4 x float> @llvm.dx.typedBufferLoad...` line entirely and still check the generated attributes, rather than having a declaration whose attributes are intentionally wrong. The only potential pitfall is that the order of the declarations might not be guaranteed - it probably depends on the traversal order of some pass or other over the IR.

For example, consider https://hlsl.godbolt.org/z/6bohv41s5 - the source doesn't declare any intrinsics, but they're all present in the output of `opt` running no passes. Bonus - you don't even have to spell out the overload parts of the intrinsic names if you don't want to.

pow2clk replied:
I've restored the tests for intrinsic attributes as you suggested.

pow2clk commented on llvm/test/CodeGen/DirectX/ResourceGlobalElimination.ll:
```diff
@@ -0,0 +1,94 @@
+!19 = !{!"int", !9, i64 0}
+!20 = !{!21}
+!21 = distinct !{!21, !22, !"_ZN4hlsl8__detail18TypedResourceProxyIU9_Res_u_CTDv4_fu17__hlsl_resource_tS2_EaSES2_: %agg.result"}
+!22 = distinct !{!22, !"_ZN4hlsl8__detail18TypedResourceProxyIU9_Res_u_CTDv4_fu17__hlsl_resource_tS2_EaSES2_"}
```
Removed. Funny that this shows as commenting on a large number of lines, but besides the text at the top, there's no visual indication of it.

pow2clk commented on llvm/include/llvm/IR/IntrinsicsDirectX.td:
```diff
@@ -28,12 +28,12 @@ def int_dx_handle_fromBinding
+                            [llvm_any_ty, llvm_i32_ty], [IntrReadMem]>;
 def int_dx_typedBufferStore
-    : DefaultAttrsIntrinsic<[], [llvm_any_ty, llvm_i32_ty, llvm_anyvector_ty]>;
+    : DefaultAttrsIntrinsic<[], [llvm_any_ty, llvm_i32_ty, llvm_anyvector_ty], [IntrWriteMem]>;
```
Done. I don't know how to apply clang-format-diff to a specific file that's not included by the existing rules

pow2clk commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -204,6 +204,22 @@ class OpLowerer {
+        Value *V = Store->getOperand(1);
+        Store->eraseFromParent();
+        if (GlobalVariable *GV = dyn_cast<GlobalVariable>(V))
+          if (GV->use_empty()) {
```
Done.

Commit:
Respond to sundry feedback

Format IntrinsicsDirectX.td

Add todo comment about future work

clean up test to remove unneeded elements and better name check prefixes

pow2clk reviewed (commented):
Responses

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into elim_res_loads_pr

Commit:
restore test for correct memory attributes applied to intrinsics

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into elim_res_loads_pr
